### ISSUE TITLE
Determining type based on previous condition

### ISSUE THREAD

**Issue by Stadly** [phpstan/phpstan-strict-rules#33]
See the following code. When evaluating `$bar+1` in the second condition, we know that `0 === $foo`. (If not, the condition would have short-circuited.) Since `0 === $foo`, `$bar` must be an integer. (If it was `null`, the function would have returned in the first condition.) Hence evaluating `$bar+1` should be fine. However, PHPStan gives the following error: `Only numeric types are allowed in +, int|null given on the left side.`.

```php
function foo(int $foo, ?int $bar): bool
{
    if (0 === $foo && null === $bar) {
        return true;
    }

    if (0 === $foo && 0 <= $bar+1) {
        return true;
    }

    return false;
}
```

**Comment by ondrejmirtes** [comment]
Hi, this is basically the same problem as https://github.com/phpstan/phpstan/milestone/12 - you should rewrite your code to be more static analyser-friendly, it's not easy to solve 😊
